cmd/snap/cmd_run.go

```diff
@@ -227,6 +231,125 @@ func isReexeced() bool {
 	return strings.HasPrefix(exe, dirs.SnapMountDir)
 }
 
+func migrateXauthority(info *snap.Info) error {
+	u, err := userCurrent()
+	if err != nil {
+		return fmt.Errorf(i18n.G("cannot get the current user: %s"), err)
+	}
+
+	// If our target directory (XDG_RUNTIME_DIR) doesn't exist we
+	// don't attempt to create it.
+	baseTargetDir := filepath.Join(dirs.XdgRuntimeDirBase, u.Uid)
+	if !osutil.FileExists(baseTargetDir) {
+		return fmt.Errorf("Target directory %s does not exist", baseTargetDir)
+	}
+
+	xauthPath := osGetenv("XAUTHORITY")
+	if len(xauthPath) == 0 || !osutil.FileExists(xauthPath) {
+		// Nothing to do for us. Most likely running outside of any
+		// graphical X11 session.
+		return nil
+	}
+
+	fin, err := os.Open(xauthPath)
+	if err != nil {
+		return err
+	}
+
+	// Abs() also calls Clean(); see https://golang.org/pkg/path/filepath/#Abs
+	xauthPathAbs, err := filepath.Abs(fin.Name())
+	if err != nil {
+		return nil
+	}
+
+	// Remove all symlinks from path
+	xauthPathCan, err := filepath.EvalSymlinks(xauthPathAbs)
+	if err != nil {
+		return nil
+	}
+
+	// Verify XAUTHORITY matches the cleaned path
+	if fin.Name() != xauthPathCan {
+		logger.Noticef("WARNING: %s != %s\n", xauthPath, xauthPathCan)
+		return nil
+	}
+
+	// Only do the migration from /tmp since /tmp is what is in a different
+	// mount namespace. There is a TOCTOU between this and the
+	// copy below, but this is just a quick check before validating the file
+	if !strings.HasPrefix(fin.Name(), "/tmp") {
+		return nil
+	}
+
+	// Ensure that the file is owned by the current user
+	fi, err := fin.Stat()
+	if err != nil {
+		return err
+	} else {
+		sys := fi.Sys()
+		if sys == nil {
+			return fmt.Errorf("Can't validate file owner")
+		}
+		// cheap comparison but better to convert uid to a string than
+		// a string into a number.
+		if fmt.Sprintf("%d", sys.(*syscall.Stat_t).Uid) != u.Uid {
+			return fmt.Errorf("Xauthority file isn't owned by the current user")
+		}
+	}
+
+	targetPath := filepath.Join(baseTargetDir, ".Xauthority")
+
+	// Only validate Xauthority file again when both files don't match
+	// otherwise we can continue using the existing Xauthority file.
+	// This is ok to do here because we aren't trying to protect against
+	// the user changing the Xauthority file in XDG_RUNTIME_DIR outside
+	// of snapd.
+	var fout *os.File
+	if osutil.FileExists(targetPath) {
+		if fout, err = os.Open(targetPath); err != nil {
+			return err
+		}
+		if osutil.FileStreamsEqual(fin, fout) {
+			osSetenv("XAUTHORITY", targetPath)
+			return nil
+		}
+	}
+
+	// To guard against setting XAUTHORITY to non-xauth files, check
+	// that we have a valid Xauthority. Specifically, the file must be
+	// parseable as an Xauthority file and not be empty.
+	if err := x11.ValidateXauthority(fin.Name()); err != nil {
+		logger.Noticef("WARNING: invalid Xauthority file: %s", err)
+		return nil
+	}
+
+	fout, err = os.OpenFile(targetPath, os.O_WRONLY|os.O_CREATE, 0600)
```

Comment:
This file needs to be securely created so it really needs os.O_EXCL, however the program logic atm means this file might exist (with different contents from XAUTHORITY). I suggest deleting the file and then using O_EXCL. Otherwise, create a temporary file somewhere else and move it into place (eg, via osutil.AtomicWriteFile).